File: src/main/java/com/leijendary/spring/boot/template/data/LocalizedData.java
package com.leijendary.spring.boot.template.data;

import static com.leijendary.spring.boot.template.util.RequestContext.getLanguage;
import static java.util.Comparator.comparingInt;
import static java.util.stream.Collectors.toList;

import java.io.Serializable;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

import lombok.Data;

@Data
public abstract class LocalizedData<T extends LocaleData> implements Serializable {

    private UUID id;
    private Set<T> translations = new HashSet<>();

    public T getTranslation() {
        final var language = getLanguage();
        final var sorted = translations
                .stream()
                .sorted(comparingInt(LocaleData::getOrdinal))
                .collect(toList());

        return sorted
                .stream()
                .filter(t -> t.getLanguage().equals(language))
                .findFirst()
                .orElse(sorted.iterator().next());
    }
}

File: src/main/kotlin/com/leijendary/spring/boot/template/core/data/LocalizedData.kt
package com.leijendary.spring.boot.template.core.data

import com.leijendary.spring.boot.template.core.util.RequestContext.language
import java.io.Serializable
import java.util.*

abstract class LocalizedData<T : LocaleData?> : Serializable {
    var id: UUID? = null
    var translations: Set<T> = HashSet()

    val translation: T
        get() {
            val language: String = language
            val sorted = translations.sortedBy { it!!.ordinal }

            return sorted.first { it!!.language == language } ?: sorted.first()
        }
}